src/main/java/com/launchdarkly/client/FeatureFlag.java

```diff
@@ -53,9 +53,19 @@ static FeatureFlag fromJson(String json) {
   }
 
   EvalResult evaluate(LDUser user, FeatureStore featureStore) throws EvaluationException {
+    if (user == null || user.getKeyAsString().isEmpty()) {
+      logger.warn("Null user or null/empty user key when evaluating flag: " + key + "; returning null");
+      return null;
+    }
     List<FeatureRequestEvent> prereqEvents = new ArrayList<>();
-    JsonElement value = evaluate(user, featureStore, prereqEvents);
-    return new EvalResult(value, prereqEvents);
+    if (isOn()) {
```

Comment:
I moved all flag eval logic into the FeatureFlag class. This enables us to call flag.evaluate() in the client's allFLags() method without additional logic around on/off behavior. Also, from an OO perspective it makes more sense: the object contains the data + the behavior.